File: c++/beeyond/PKCS12PBEKey.cxx
#define BEECRYPT_CXX_DLL_EXPORT

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include "beecrypt/c++/beeyond/PKCS12PBEKey.h"

using namespace beecrypt::beeyond;

PKCS12PBEKey::PKCS12PBEKey(const array<jchar>& password, const bytearray* salt, int iterationCount) : _pswd(password)
{
	if (salt)
		_salt = new bytearray(*salt);
	else
		_salt = 0;
	_iter = iterationCount;
	_enc = 0;
}

PKCS12PBEKey::~PKCS12PBEKey()
{
	delete _salt;
}

bool PKCS12PBEKey::operator==(const Key& compare) const throw ()
{
	const PBEKey* tmp = dynamic_cast<const PBEKey*>(&compare);
	if (tmp)
	{
		if (_pswd != tmp->getPassword())
			return false;

		if (_salt && tmp->getSalt())
		{
			if (*_salt != *tmp->getSalt())
				return false;
		}
		else if (_salt || tmp->getSalt())
			return false;
				
		if (_iter != tmp->getIterationCount())
			return false;

		return true;
	}
	return false;
}

PKCS12PBEKey* PKCS12PBEKey::clone() const
{
	return new PKCS12PBEKey(_pswd, _salt, _iter);
}

bytearray* PKCS12PBEKey::encode(const array<jchar>& password)
{
	int i;

	bytearray* result = new bytearray((password.size() + 1) * 2);

	for (i = 0; i < password.size(); i++)
	{
		(*result)[2*i  ] = (password[i] >> 8) & 0xff;
		(*result)[2*i+1] = (password[i]     ) & 0xff;
	}
	(*result)[2*i  ] = 0;
	(*result)[2*i+1] = 0;

	return result;
}

int PKCS12PBEKey::getIterationCount() const throw ()
{
	return _iter;
}

const array<jchar>& PKCS12PBEKey::getPassword() const throw ()
{
	return _pswd;
}

const bytearray* PKCS12PBEKey::getSalt() const throw ()
{
	return _salt;
}

const bytearray* PKCS12PBEKey::getEncoded() const throw ()
{
	if (!_enc)
		_enc = encode(_pswd);

	return _enc;
}

const String& PKCS12PBEKey::getAlgorithm() const throw ()
{
	static const String ALGORITHM("PKCS#12/PBE");
	return ALGORITHM;
}

const String* PKCS12PBEKey::getFormat() const throw ()
{
	static const String FORMAT("RAW");
	return &FORMAT;
}

File: java/src/beecrypt/beeyond/PKCS12PBEKey.java
package beecrypt.beeyond;

import javax.crypto.interfaces.*;

public class PKCS12PBEKey implements PBEKey {
	private char[] _pwd;
	private byte[] _salt;
	private int _iter;
	private byte[] _enc;

	public static byte[] encode(char[] password) {
		int i;
		byte[] result = new byte[(password.length + 1) * 2];
		for (i = 0; i < password.length; i++) {
			char c = password[i];
			result[2 * i] = (byte) (c >> 8);
			result[2 * i + 1] = (byte) c;
		}
		result[2 * i] = 0;
		result[2 * i + 1] = 0;
		return result;
	}

	public PKCS12PBEKey(char[] password, byte[] salt, int iterationCount) {
		_pwd = password;
		_salt = salt;
		_iter = iterationCount;
	}

	public String getAlgorithm() {
		return "PKCS#12/PBE";
	}

	public byte[] getEncoded() {
		if (_enc == null)
			_enc = encode(_pwd);

		return _enc;
	}

	public String getFormat() {
		return "RAW";
	}

	public int getIterationCount() {
		return _iter;
	}

	public char[] getPassword() {
		return _pwd;
	}

	public byte[] getSalt() {
		return _salt;
	}

}
